src/validation/validation.cpp

```diff
@@ -2474,32 +2474,34 @@ bool ConnectBlockCanonicalOrdering(const CBlock &block,
 
             if (!tx.IsCoinBase())
             {
-                if (!view.HaveInputs(tx))
-                {
-                    // If we were validating at the same time as another block and the other block wins the validation
-                    // race
-                    // and updates the UTXO first, then we may end up here with missing inputs.  Therefore we checke to
-                    // see
-                    // if the chainwork has advanced or if we recieved a quit and if so return without DOSing the node.
-                    if (PV->ChainWorkHasChanged(nStartingChainWork) || PV->QuitReceived(this_id, fParallel))
-                    {
-                        return false;
-                    }
-                    return state.DoS(100, error("%s: block %s inputs missing/spent in tx %d %s", __func__,
-                                              block.GetHash().ToString(), i, tx.GetHash().ToString()),
-                        REJECT_INVALID, "bad-txns-inputs-missingorspent");
-                }
-
                 // Check that transaction is BIP68 final
                 // BIP68 lock checks (as opposed to nLockTime checks) must
                 // be in ConnectBlock because they require the UTXO set
                 prevheights.resize(tx.vin.size());
                 {
                     for (size_t j = 0; j < tx.vin.size(); j++)
                     {
-                        prevheights[j] = CoinAccessor(view, tx.vin[j].prevout)->nHeight;
+                        CoinAccessor coin(view, tx.vin[j].prevout);
+                        // isSpend is true for empty coin object (coinEmpty)
+                        if (coin->IsSpent())
+                        {
```

Comment:
It's not entirely clear to me what happens here if coin returns with a coin not found...or returns with &emptyCoin.... will that trigger coin->IsSpent?  coin->IsSpent() will be true if out.SetNull() is true but I'm not sure if that's the case here.

Reply:
CoinAccessor returns either the found coin or emptyCoin. emptyCoin->IsSpent() is always true.

emptyCoin calls the default constructor for CTxOut when it creates its out datamember, and the default constructor for CTxOut calls SetNull() so IsSpent for emptyCoin should always be true.

Reply:
yes that's right...ok good. will approve now.